### ISSUE TITLE
Parameter wrapping should take priority over expression wrapping

### ISSUE THREAD

**Issue by mklnwt** [pinterest/ktlint#2454]
`function_signature_rule_force_multiline_when_parameter_count_greater_or_equal_than` should take priority over `function_signature_body_expression_wrapping`.

This can save horizontal space at the cost of 1 line vertical space.

```Kotlin
// Assume that the last allowed character is
// at the X character on the right                  X
fun foo(bar: String = "loooooooooong"): Foo = Foo.baz(a = "looooooooooooooooooooooooooooooooooooong", b = "looooooooooooooooooooooooooooooooooooong")
```

## Expected Behavior
```Kotlin
// Assume that the last allowed character is
// at the X character on the right                  X
fun foo(
    bar: String = "loooooooooong",
): Foo = Foo.baz(
    a = "looooooooooooooooooooooooooooooooooooong",
    b = "looooooooooooooooooooooooooooooooooooong",
)
```

## Current Behavior
```Kotlin
// Assume that the last allowed character is
// at the X character on the right                  X
fun foo(bar: String = "loooooooooong"): Foo =
    Foo.baz(
        a = "looooooooooooooooooooooooooooooooooooong",
        b = "looooooooooooooooooooooooooooooooooooong",
    )
```

Additional error:
> Exceeded max line length (?) (cannot be auto-corrected) (standard:max-line-length)

## Additional information
* Current version of ktlint: 1.1.0

**Comment by paul-dingemans** [comment]
This can be achieved by setting `ktlint_function_signature_body_expression_wrapping`. See https://pinterest.github.io/ktlint/latest/rules/configuration-ktlint/#wrapping-the-expression-body-of-a-function

**Comment by mklnwt** [comment]
No setting of `ktlint_function_signature_body_expression_wrapping` achieves the expected result.

**Comment by paul-dingemans** [comment]
Let me investigate this once more.

**Comment by paul-dingemans** [comment]
Can you double check that you are using the configuration settings correctly? I see in this post several references to settings which are missing the `ktlint_` prefix.

Given `.editorconfig`:
```
root = true

[*.{kt,kts}]
ktlint_code_style = ktlint_official
ktlint_function_signature_rule_force_multiline_when_parameter_count_greater_or_equal_than = 1
ktlint_function_signature_body_expression_wrapping = default
max_line_length = 53
ktlint_standard = enabled
ktlint_experimental = enabled
```
the code below:
```
// Assume that the last allowed character is
// at the X character on the right                  X
fun foo(bar: String = "loooooooooong"): Foo = Foo.baz(
    a = "looooooooooooooooooooooooooooooooooooong",
    b = "looooooooooooooooooooooooooooooooooooong",
)
```
is formatted as:
```
// Assume that the last allowed character is
// at the X character on the right                  X
fun foo(
    bar: String = "loooooooooong",
): Foo = Foo.baz(
    a = "looooooooooooooooooooooooooooooooooooong",
    b = "looooooooooooooooooooooooooooooooooooong",
)
```

**Comment by mklnwt** [comment]
I have not set `ktlint_function_signature_rule_force_multiline_when_parameter_count_greater_or_equal_than = 1` and left it with the default of 2.

**Comment by paul-dingemans** [comment]
On response to my suggestion:
> This can be achieved by setting ktlint_function_signature_body_expression_wrapping. See https://pinterest.github.io/ktlint/latest/rules/configuration-ktlint/#wrapping-the-expression-body-of-a-function

you said that:
> No setting of ktlint_function_signature_body_expression_wrapping achieves the expected result.

You can achieve what you want, by setting it to 1.

**Comment by mklnwt** [comment]
I'm trying to say that it should be wrapped based on line length, not because of the parameter number.

With the default setting (2) both of the following should be correctly formatted.

```Kotlin
// Assume that the last allowed character is
// at the X character on the right                  X
fun foo(
    bar: String = "loooooooooong",
): Foo = Foo.baz(
    a = "looooooooooooooooooooooooooooooooooooong",
    b = "looooooooooooooooooooooooooooooooooooong",
)
```

```Kotlin
// Assume that the last allowed character is
// at the X character on the right                  X
fun foo(bar: String = "short"): Foo = Foo.baz(
    a = "looooooooooooooooooooooooooooooooooooong",
    b = "looooooooooooooooooooooooooooooooooooong",
)
```

**Comment by paul-dingemans** [comment]
Please provide your `.editorconfig` settings.

When I use:
```
root = true

[*.{kt,kts}]
ktlint_code_style = intellij_idea
max_line_length = 53
ktlint_standard = enabled
ktlint_experimental = enabled
```
code is formatted as you requested. Note that `max_line_length` must be set explicitly when using code style `intellij_idea`.

**Comment by mklnwt** [comment]
Minimal example `.editorconfig`:

```
root = true

[*.{kt,kts}]
max_line_length = 53
ktlint_code_style = ktlint_official
ktlint_standard = enabled
ktlint_experimental = enabled
```

**Comment by paul-dingemans** [comment]
You need to overwrite `ktlint_function_signature_body_expression_wrapping`.

```
root = true

[*.{kt,kts}]
max_line_length = 53
ktlint_code_style = ktlint_official
ktlint_standard = enabled
ktlint_experimental = enabled
ktlint_function_signature_body_expression_wrapping = default
```

**Comment by mklnwt** [comment]
Actually while producing a minimal example I might've accidentally omitted that part and also forgot another crucial overwrite,
Here is a complete picture that I'm now running on `1.1.1`:
```
[*.{kt,kts}]
max_line_length = 53
ktlint_standard = enabled
ktlint_experimental = enabled
ktlint_code_style = ktlint_official
ktlint_function_signature_body_expression_wrapping = default
ktlint_standard_multiline-expression-wrapping = disabled
ktlint_standard_string-template-indent = disabled # depends on multiline-expression-wrapping
```

Still, my examples show violations:

```Kotlin
// Assume that the last allowed character is
// at the X character on the right                  X
fun foo(
    bar: String = "loooooooooong",
): Foo = Foo.baz(
    a = "looooooooooooooooooooooooooooooooooooong",
    b = "looooooooooooooooooooooooooooooooooooong",
)
```

> 4:5 No whitespace expected between opening parenthesis and first parameter name (standard:function-signature)
> 4:35 No whitespace expected between last parameter and closing parenthesis (standard:function-signature)
> 5:10 Newline expected before expression body (standard:function-signature)

**Comment by paul-dingemans** [comment]
I still can not reproduce this. With `.editorconfig` and code sample above, I get following:
```
17:56:26.944 [main] INFO com.pinterest.ktlint.cli.internal.KtlintCommandLine -- Enable default patterns [**/*.kt, **/*.kts]
src/main/kotlin/Foo.kt:5:54: Exceeded max line length (53) (standard:max-line-length)
src/main/kotlin/Foo.kt:6:54: Exceeded max line length (53) (standard:max-line-length)

Summary error count (descending) by rule:
  standard:max-line-length: 2
```

But, I do note that your `.editorconfig` is missing line below which should be at the top of the file:
```
root = true
```
Without this property, you might actually pick-up an `.editorconfig` file at a higher level in your directory structure.

**Comment by mklnwt** [comment]
I tried it with and without root.

Did you run check or format?
With format I get the same error as you, but the code is reformatted as stated above.

```kotlin
fun foo(bar: String = "loooooooooong"): Foo =
    Foo.baz(
        a = "looooooooooooooooooooooooooooooooooooong",
        b = "looooooooooooooooooooooooooooooooooooong",
    )
```

**Comment by paul-dingemans** [comment]
> With format I get the same error as you, but the code is reformatted as stated above.

Did you mean "below" instead of "above"?

Please try to be more explicit, because I getting confused and need to guess. So please, specify:
* (full) editorconfig
* Input
* Output
* Expected output

**Comment by mklnwt** [comment]
Sorry for the confusion. I wrote tests instead.

If `ktlint_function_signature_body_expression_wrapping` is set to `multiline` or `always`, everything works as expected.
If it is set to `default`, `argument-list-wrapping` should wrap regardless of the value of `ktlint_function_signature_rule_force_multiline_when_parameter_count_greater_or_equal_than`, because the remainder after `=` doesn't fit into the line.

After trying a few more examples, though, I think my expected cases would cause too much complexity. For this to work `argument-list-wrapping` & `function_signature` would have to interact with each other.

I think it's better to leave the issue closed, but for completeness sake and future reference I'll leave the expectation and test cases here.

Ideally both of the following would be allowed:

```kotlin
fun foo(
    bar: String
): Foo = Foo.baz(
    a = "looooooooooooooooooooong",
    b = "looooooooooooooooooooong"
)

fun foo(bar: String): String =
    Foo.baz(a = "a", b = "b")
```

Test cases

```kotlin
class ExampleTest {
    private val argumentListWrappingRuleAssertThat =
        assertThatRuleBuilder { ArgumentListWrappingRule() }
            .addRequiredRuleProviderDependenciesFrom(StandardRuleSetProvider())
            .assertThat()

    @Test
    fun `Expected signature_body_expression_wrapping=default (unformatted)`() {
        val code =
            """
            fun foo(bar: String): Foo = Foo.baz(a = "looooooooooooooooooooong", b = "looooooooooooooooooooong")
            """.trimIndent()
        val formattedCode =
            """
            fun foo(
                bar: String
            ): Foo = Foo.baz(
                a = "looooooooooooooooooooong",
                b = "looooooooooooooooooooong"
            )
            """.trimIndent()
        argumentListWrappingRuleAssertThat(code)
            .addAdditionalRuleProviders(
                { FunctionSignatureRule() },
                { IndentationRule() },
            ).withEditorConfigOverride(MAX_LINE_LENGTH_PROPERTY to 35)
            .withEditorConfigOverride(FUNCTION_BODY_EXPRESSION_WRAPPING_PROPERTY to "default")
            .isFormattedAs(formattedCode)
    }

    @Test
    fun `Expected signature_body_expression_wrapping=default (pre-formatted)`() {
        val code =
            """
            fun foo(
                bar: String
            ): Foo = Foo.baz(
                a = "looooooooooooooooooooong",
                b = "looooooooooooooooooooong"
            )
            """.trimIndent()
        argumentListWrappingRuleAssertThat(code)
            .addAdditionalRuleProviders(
                { FunctionSignatureRule() },
                { IndentationRule() },
            ).withEditorConfigOverride(MAX_LINE_LENGTH_PROPERTY to 35)
            .withEditorConfigOverride(FUNCTION_BODY_EXPRESSION_WRAPPING_PROPERTY to "default")
            .hasNoLintViolations()
    }

    @Test
    fun `Current signature_body_expression_wrapping=default (pre-formatted)`() {
        val code =
            """
            fun foo(
                bar: String
            ): Foo = Foo.baz(
                a = "looooooooooooooooooooong",
                b = "looooooooooooooooooooong"
            )
            """.trimIndent()
        val formattedCode =
            """
            fun foo(bar: String): Foo =
                Foo.baz(
                    a = "looooooooooooooooooooong",
                    b = "looooooooooooooooooooong"
                )
            """.trimIndent()
        argumentListWrappingRuleAssertThat(code)
            .addAdditionalRuleProviders(
                { FunctionSignatureRule() },
                { IndentationRule() },
            ).withEditorConfigOverride(MAX_LINE_LENGTH_PROPERTY to 35)
            .withEditorConfigOverride(FUNCTION_BODY_EXPRESSION_WRAPPING_PROPERTY to "default")
            .hasLintViolationsForAdditionalRule(
                LintViolation(2, 5, "No whitespace expected between opening parenthesis and first parameter name", true),
                LintViolation(2, 16, "No whitespace expected between last parameter and closing parenthesis", true),
                LintViolation(3, 10, "Newline expected before expression body", true),
            ).isFormattedAs(formattedCode)
    }

    @Test
    fun `Current & Expected signature_body_expression_wrapping=multiline`() {
        val code =
            """
            fun foo(
                bar: String
            ): Foo = Foo.baz(
                a = "looooooooooooooooooooong",
                b = "looooooooooooooooooooong"
            )
            """.trimIndent()
        val formattedCode =
            """
            fun foo(bar: String): Foo =
                Foo.baz(
                    a = "looooooooooooooooooooong",
                    b = "looooooooooooooooooooong"
                )
            """.trimIndent()
        argumentListWrappingRuleAssertThat(code)
            .addAdditionalRuleProviders(
                { FunctionSignatureRule() },
                { IndentationRule() },
            ).withEditorConfigOverride(MAX_LINE_LENGTH_PROPERTY to 35)
            .withEditorConfigOverride(FUNCTION_BODY_EXPRESSION_WRAPPING_PROPERTY to "multiline")
            .hasLintViolationsForAdditionalRule(
                LintViolation(2, 5, "No whitespace expected between opening parenthesis and first parameter name", true),
                LintViolation(2, 16, "No whitespace expected between last parameter and closing parenthesis", true),
                LintViolation(3, 10, "Newline expected before expression body", true),
            ).isFormattedAs(formattedCode)
    }

    @Test
    fun `Current & Expected where argument-list-wrapping would lead to additional breaks signature_body_expression_wrapping=default (pre-formatted)`() {
        val code =
            """
            fun foo(bar: String): String =
                Foo.baz(a = "a", b = "b")
            """.trimIndent()
        argumentListWrappingRuleAssertThat(code)
            .addAdditionalRuleProviders(
                { FunctionSignatureRule() },
                { IndentationRule() },
            ).withEditorConfigOverride(MAX_LINE_LENGTH_PROPERTY to 35)
            .withEditorConfigOverride(FUNCTION_BODY_EXPRESSION_WRAPPING_PROPERTY to "default")
            .hasNoLintViolations()
    }
}
```

**Comment by paul-dingemans** [comment]
It took a while before I had a chance to look at this again. It is not that difficult to achieve now you already provided as set of tests that would statisfy you. The `argument-list-wrapping` rule already had a dependency on the `class-signature` rule for more or less the same reason. This has now been added for the `function-signature` rule as well.
